crates/fortitude_preprocessor/src/cpp_tokens.rs

```diff
@@ -0,0 +1,618 @@
+                self.step();
+                // Handle CRLF
+                if self.iter.peek() == Some(&b'\n') {
+                    self.step();
+                }
+                Some(self.emit(start, CppTokenKind::Newline))
+            }
+            _ => None,
+        }
+    }
+
+    /// Consumes any amount of whitespace and combinations of tabs and spaces.
+    /// Does not include newlines.
+    pub fn consume_whitespace(&mut self) -> Option<CppTokenResult<'a>> {
+        let start = self.offset;
+        while matches!(self.iter.peek(), Some(&b' ') | Some(&b'\t')) {
+            self.step();
+        }
+        if self.offset > start {
+            Some(self.emit(start, CppTokenKind::Whitespace))
+        } else {
+            None
+        }
+    }
+
+    /// Consume a comment until the end of the line.
+    fn consume_comment(&mut self) -> Option<CppTokenResult<'a>> {
+        if *self.iter.peek()? == b'!' {
+            let start = self.offset;
+            self.skip_to_line_end();
+            Some(self.emit(start, CppTokenKind::Comment))
+        } else {
+            None
+        }
+    }
+
+    /// If the next token is a string, consume it. Otherwise returns `None`.
+    /// Handles both single and double quoted strings.
+    /// Handles multiline strings with line continuations.
+    /// Does not handle escaped quotes, instead treating them as
+    /// string delimiters.
+    fn consume_string(&mut self) -> Option<CppTokenResult<'a>> {
+        let delimiter = match self.iter.peek() {
+            Some(&b'\'') => b'\'',
+            Some(&b'\"') => b'\"',
+            _ => return None,
+        };
+        let start = self.offset;
+        self.step();
+        while self.iter.peek() != Some(&delimiter) {
+            if self.iter.peek().is_none() {
+                // Unterminated string
+                return Some(self.emit(start, CppTokenKind::Error));
+            }
+            self.step();
+        }
+        self.step();
+        Some(self.emit(start, CppTokenKind::String))
+    }
+
+    /// Consume a preprocessor directive, including the leading `#`.  Any amount
+    /// of whitespace can be between the `#` and the directive name.  Also
+    /// captures stringification within macros.
+    pub fn consume_directive(&mut self) -> Option<CppTokenResult<'a>> {
+        if self.iter.peek() != Some(&b'#') {
+            return None;
+        }
+        let start = self.offset;
+        self.step();
+        // If the next character is another '#', it's a concatenation operator.
+        if self.iter.peek() == Some(&b'#') {
+            self.step();
+            return Some(self.emit(start, CppTokenKind::Concatenation));
+        }
+        // Consume any whitespace after the '#'.
+        let _ = self.consume_whitespace();
+        // Consume an identifier for the directive name.
+        match self.consume_identifier() {
+            Some(Ok(directive)) => {
+                let directive_kind = match directive.text {
```

Comment:
I think you can replace this with `#[derive(strum_macros::EnumStr)]` on `CppDirectiveKind`, using `#[strum(default)]` on `Stringification`

Reply:
Done in https://github.com/PlasmaFAIR/fortitude/pull/510/commits/3fe737b2ffebf3cb31dc0d6821d3a894b86eef01 :+1: